intelmq/lib/datatypes.py

```diff
@@ -54,8 +58,85 @@ def toJson(self):
 
 
 class LogLevel(Enum):
-    DEBUG    = 0
-    INFO     = 1
-    WARNING  = 2
-    ERROR    = 3
+    DEBUG = 0
+    INFO = 1
+    WARNING = 2
+    ERROR = 3
     CRITICAL = 4
+
+
+class TimeFormat(str):
+    """
+    Pydantic style Field Type class for bot parameter time_format. Used for validation.
+    """
+
+    def __new__(cls, value: Optional[str] = None):
+        value = "fuzzy" if (value is None or value == "") else value
```

Comment:
> Doesn't it the same as `value = value or "fuzzy"`, assuming the type is correct?

Yes you are correct, this is just more explicit. But `value = value or "fuzzy"` is more elegant definitely.

> BTW, do we need bot `__new__` and `__init__` in this case? Why not just do it in the `__init__`? There is no alternate path with a different constructor, also I don't see any reason to use `__new__`.

Because `str` is immutable and I want to manipulate it, it must be done before the object is instantiated. Therefore I need to use the `__new__` method. The manipulation is kind of a defensive measure when for example intelmq-manager badly interprets empty parameter. No need to fail/raise exception, just use the default value. This couldn't be done in the `__init__` as far as I know and from what I tried.

However adding the `self.convert` and `self.format_string` should be done in the `__init__` method. This is also where the validation is done.

This is why both `__new__` and `__init__` is used.

Reply:
I agree `value = value or "fuzzy"` is nicer and quicker to understand.

Can you please add this explanation of `__new__` into an inline comment?